### PR TITLE
Feature: print email

### LINKED ISSUE #579: Print stylesheet [$65]
Occasionally a user needs to print an email. CTRL+P will print the page, not the message. Adding a div to the messages-list-item template and some js tied to an action will not address CTRL+P but im not sure if everyone would want to override the browser shortcut anyway.  

There are some tricks in CSS To make the output in the mail-message-container cleaner ( reply attachment and forward buttons for example ) for printing but putting the print button in the message list item would at least prevent that particular control from being printed. 

Alternatively I suppose it would be even better in some cases to pull the message from the mail server again for printing purposes to avoid printing all the extraneous information mentioned above, and in the event someone entered text into the reply but had not sent it yet -- we might not want to print message text that has never been sent. 

Other considerations are whether we want the print preview dialog allowing the user to select a printer or to blindly print to whatever is default.   

Any thoughts on this?  I can probably contribute to this one if its not already in progress or part of NCMail13 but Id like to hear some input from others. 

thx
-d

<bountysource-plugin>

---
There is a **[$65 open bounty](https://www.bountysource.com/issues/50806693-print-stylesheet?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by sourenaraya**
Fixes #579

### CODE REVIEW

File: src/components/MessagePrint.vue [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	},
+	methods: {
+		getIframeDoc() {
+			const iframe = this.$refs.iframe
+			return iframe.contentDocument || iframe.contentWindow.document
+		},
+		replaceBody() {
+			const iframeDoc = this.getIframeDoc()
+			const iframeBody = iframeDoc.querySelectorAll('body')[0]
+			console.log({header: this.header, body: this.body})
```
Comment (ChristophWurst) [inline src/components/MessagePrint.vue]: debug statement? can we remove this?

Comment (sourenaraya) [reply src/components/MessagePrint.vue]: sorry for this, removed.

**Comment by jancborchardt** [comment]
So I was initially confused because using the browser print method (via the menu or Ctrl-P) still printed the rest of the mail.

Only then I saw the "Print" button on the bottom. ;)

As mentioned in the original issue at https://github.com/nextcloud/mail/issues/579 a print stylesheet just hiding elements (like the header, `#app-navigation`, and the reply form) via CSS would be the preferred method here. It is much simpler, also works with the native print function, and we don’t introduce more controls/buttons to which are rarely used.

A good article on how to create a print stylesheet is at Smashing Magazine: https://www.smashingmagazine.com/2018/05/print-stylesheets-in-2018/

**Comment by sourenaraya** [comment]
I tried this aproach, it worked fine for plaintext emails, but i dont think its possible to print HTML emails that way. The problem is, if height of email is higher than screen height, its not being printed correctly, because HTML emails is in iframe. Setting height of iframe and all parent elements to 100% doesnt help. Not sure if its possible to implement CSS only solution, but will try again.

**Comment by sourenaraya** [comment]
Ok, i did some research. 
There is no CSS-only solution.
Using `beforeprint` event we could adjust size of iframe to its content height and then `afterprint` to remove explicit size, but safari doesnt support that, so we have to use polyfill.

**Comment by jancborchardt** [comment]
Ah ok – good research @sourenaraya! So with the polyfill would it be possible to also catch the native browser print and Ctrl-P? Cause that would be the preferred method over a dedicated "Print" button. :)

**Comment by sourenaraya** [comment]
i think so. Working on it. Do i just force-push new changes, or better close this PR and open a new one?

**Comment by jancborchardt** [comment]
@sourenaraya cool, thanks so much! Regarding force push or new pull request – feel free to do whatever is better for you. :)

**Comment by ChristophWurst** [comment]
> i think so. Working on it. Do i just force-push new changes, or better close this PR and open a new one?

https://dev.to/tmr232/dont-amend-fix :rocket:

**Comment by sourenaraya** [comment]
see #1804

**Pull request closed without merging**